packages/graph/hash_graph/lib/graph/src/api/rest/middleware.rs

```diff
@@ -0,0 +1,73 @@
+use axum::{
+    body::{Body, Bytes, HttpBody},
+    http::{Request, StatusCode},
+    middleware::Next,
+    response::{IntoResponse, Response},
+};
+use tracing::{enabled, Level};
+
+// *Heavily* inspired by
+// https://github.com/tokio-rs/axum/blob/main/examples/print-request-response/src/main.rs
+
+/// A development-environment-focused Axum Handler function to buffer the body of requests and
+/// responses and log them. Overhead should be minimal when not in `Level::Trace`
+pub(super) async fn log_request_and_response(
+    request: Request<Body>,
+    next: Next<Body>,
+) -> Result<impl IntoResponse, (StatusCode, String)> {
+    let mut request = request;
+    if enabled!(Level::TRACE) {
+        // Destructure the request and pass the stream buffer
+        let (parts, body) = request.into_parts();
+        let bytes = buffer_and_log("request", body, None).await?;
+        request = Request::from_parts(parts, Body::from(bytes));
+    }
+
+    // Run the rest of the layers
+    let response = next.run(request).await;
+
+    if enabled!(Level::TRACE) {
```

Comment:
This could potentially be a lot of data. I think we should maybe make this an additional environment variable or maybe feature flag.

Reply:
cc @thehabbos007 @TimDiekmann

Reply:
Especially since we don't have non-blocking logging atm

Reply:
I find it useful to be able to switch the logging on-the-fly. Does the logging still take place but gets filtered out just before flushing, or how does the level affect the output?

Reply:
I think it's worth it for us to do what Alfie said above having another environment variable that enables body logging in trace mode, essentially making it so we have a 
```rust
    if enabled!(Level::TRACE) && env_set!("LOG_HTTP_BODY") { ..
```
such that we don't run the code unless we really want to log the HTTP bodies - even in TRACE mode.

Having looked at the `enabled!` implementation, I don't think we should go about implementing it right now. If we can find a crate to do this, we could use that right now, though.